File: algorithms/coding-interview-questions-blind75-lol/1/floodfill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        
        n = len(image)
        m = len(image[0])
        init_color = image[sr][sc]


        def dfs(i, j):
            image[i][j] = color
            if i + 1 < n and image[i+1][j] == init_color:
                dfs(i+1, j)
            if j + 1 < m and image[i][j+1] == init_color:
                dfs(i, j+1)
            if i > 0 and image[i-1][j] == init_color:
                dfs(i-1, j)
            if j > 0 and image[i][j-1] == init_color:
                dfs(i, j-1)
        
        if image[sr][sc] != color:
            dfs(sr, sc)

        return image

File: algorithms/blind75/1/floodfill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {

        if (image[sr][sc] != color)
            dfs(image, image[sr][sc], sr, sc, color, size(image), size(image[0]));

        return image;
    }

    void dfs(vector<vector<int>> & image, int startColor, int i, int j, int color, int n, int m) {

        image[i][j] = color;

        if (i+1 < n && image[i+1][j] == startColor)
            dfs(image, startColor, i+1, j, color, n, m);
        if (j+1 < m && image[i][j+1] == startColor)
            dfs(image, startColor, i, j+1, color, n, m);
        if (i-1 >= 0 && image[i-1][j] == startColor)
            dfs(image, startColor, i-1, j, color, n, m);
        if (j-1 >= 0 && image[i][j-1] == startColor)
            dfs(image, startColor, i, j-1, color, n, m);
    }
};
